### PR TITLE
reopening transfer feature PR

### LINKED ISSUE #868: Transfer question Context Command
## Motivation

New users frequently post questions in `#chit_chat` and it is impossible to educate them all. On top, when told to post in `#questions`, users often just dont post anything at all anymore.

## Feature

To ease above, we should add a message context command that allows staff to transfer a wrongly posted question into the right channel. The bot then does all the heavy lifting and educates the user via DM.

I.e. staff right clicks a wrongly posted message and selects a new action `transfer-question`

transfer

The bot then opens a modal window where the staff can select a title, the message and tags for the post. The bot can prefill the message with the users original message and the title can be prefilled with a guess (see existing logic for `/suggestion`, which has that as well). The tag can be prefilled with just `Java` - a dropdown for the tags would be totally fine.

The staff then clicks `OK` to submit the modal. The bot creates the post in `#questions` as embed (setting the original user as author of the embed). The bot deletes the original message and DMs the original user about what happened. Sth along the lines of:

> Hey there, unfortunately you have asked your question in the wrong channel (see `#welcome` for details). Your question was automatically transferred to `#link_to_the_post`. Please continue there, thanks 👍

If the DM fails, the bot sends the message in the original channel (`#chit_chat`) instead.

## Details

It is of course totally okay to add the feature piece by piece and just start with a quick draft first, instead of doing all features.

### PR DESCRIPTION
**Pull request by ankitsmt211**
Adds transfer feature
* implemented suggested changes
* fixes #868

### CODE REVIEW

**Comment by ankitsmt211** [comment]
i believe the it's failing because of 
image

**Comment by Zabuzard** [comment]
rebase please

**Comment by ankitsmt211** [comment]
Linking screenshots from the original PR, all the steps are intact except step 7 and no more pings when linking new post in source channel, only username is mentioned(since a DM and a mention in new post is already done).
 https://github.com/Together-Java/TJ-Bot/pull/885#issuecomment-1723088379

**Review by Taz03 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/features/moderation/TransferQuestionCommand.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        List<ForumTag> tags = questionsForum.getAvailableTagsByName("JAVA", true);
+        ForumTag javaTag = tags.get(0);
+
+        questionsForum.createForumPost(forumTitle, forumMessage)
+            .setTags(ForumTagSnowflake.fromId(javaTag.getId()))
+            .map(ForumPost::getMessage)
+            .onSuccess(message -> message.reply("original post from" + originalUser.getAsMention())
+                .queue(onSuccess -> {
+                }, onFailure -> logger.warn("Failed to mention user in transferred post",
+                        onFailure)))
```
Comment (Taz03) [inline application/src/main/java/org/togetherjava/tjbot/features/moderation/TransferQuestionCommand.java:135]: don't log here

**Review by Taz03 — CHANGES_REQUESTED** [review]

**Comment by ankitsmt211** [comment]
I'm thinking to refactor the process, 
Earlier
* Create post
* Mention OP
* DM OP
* link to new post with OP's name in source channel

Suggested
* Create Post
* Mention OP in post
* DM OP

Reasons: Since Op is mentioned twice during the process, they are already redirected to new post. There's no benefit to posting link to new post apart from possibily adding unwanted text source channel along with promotion of newly created post.

**Comment by Taz03** [comment]
Agree, but i'd say give the creator an ephemeral message in source channel linking the forum post

**Comment by ankitsmt211** [comment]
> Agree, but i'd say give the creator an ephemeral message in source channel linking the forum post

Seems like ephermal messages can be only sent to interactions, but OP is not the one interacting. So any ideas with that?

**Comment by ankitsmt211** [comment]
> Agree, but i'd say give the creator an ephemeral message in source channel linking the forum post

Do you mean maybe whoever is doing the transfer? Since that makes sense if they were able to make the transfer complete?

**Comment by marko-radosavljevic** [comment]
> I'm thinking to refactor the process

Agree, will copy my response from previous PR:
> Not a fan of "three" back to back mentions are being triggered here for OP
> 
>     * on creation of new post
> 
>     * private DM
> 
>     * link to newly created post in source channel

I agree, let's get rid of the mention in source channel.

That way we don't annoy and confuse the newcomer with mentions all over the place. They only get mentioned where we want them to go, to a correct help thread. Additionally, we keep the source channel clean, and don't interrupt ongoing discussions.

Seems like a cleaner and more streamlined approach. :relaxed:

**Comment by marko-radosavljevic** [comment]
> Seems like ephermal messages can be only sent to interactions, but OP is not the one interacting. So any ideas with that?

Yeah, as far as I know, it's not possible to do something like that.

> Do you mean maybe whoever is doing the transfer? Since that makes sense if they were able to make the transfer complete?

An ephemeral message to the person doing the transfer is probably not needed, they already know what happened and why. There isn't much to notify them about, either.

**Comment by ankitsmt211** [comment]
I was able to squeeze all of it into "one" chain, and make functional. Failed DM's are handled inside `dmUser` using `ErrorFlatMap`.
Apart from that, should i log anything else from this chain? 

  ```java  
@Override
    public void onModalSubmitted(ModalInteractionEvent event, List<String> args) {
        event.getJDA()
            .retrieveUserById(args.get(0))
          .map(user->user)
          .flatMap(fetchedUser->createForumPost(event, fetchedUser))
          .flatMap(user->dmUser(event.getChannel(), user))
          .flatMap(deleted->event.reply("transferred to"+deleted.getAsMention()).setEphemeral(true))
          .flatMap(info->deleteOriginalMessage(event.getJDA(),args.get(2),args.get(1)))
            .queue();
    }
```
>  .flatMap(deleted->event.reply("transferred to"+deleted.getAsMention()).setEphemeral(true))

Needed to reply to event, otherwise modal menu wouldn't close and showed error at the top. 

if this is good to go, i can commit and push for any further tweaks.

**Pull request closed without merging**